ostd/src/sync/rcu/mod.rs

```diff
@@ -394,32 +398,51 @@ impl<P: NonNullPtr + Send> RcuOptionReadGuard<'_, P> {
     }
 }
 
-/// # Safety
+/// A wrapper to delay calling destructor of `T` after the RCU grace period.
+///
+/// Upon dropping this structure, a callback will be registered to the global
+/// RCU monitor. And the destructor of `T` will not be immediately called.
 ///
-/// The pointer must be previously returned by `into_raw` and the pointer
-/// must be only be dropped once.
-unsafe fn delay_drop<P: NonNullPtr + Send>(pointer: NonNull<<P as NonNullPtr>::Target>) {
-    struct ForceSend<P: NonNullPtr + Send>(NonNull<<P as NonNullPtr>::Target>);
-    // SAFETY: Sending a raw pointer to another task is safe as long as
-    // the pointer access in another task is safe (guaranteed by the trait
-    // bound `P: Send`).
-    unsafe impl<P: NonNullPtr + Send> Send for ForceSend<P> {}
+/// `RcuDrop<T>` is guaranteed to have the same layout as `T`. You can also
+/// access the inner value safely via `RcuDrop<T>`.
+#[derive(Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[repr(transparent)]
+pub struct RcuDrop<T: Send + 'static> {
+    value: ManuallyDrop<T>,
+}
 
-    let pointer: ForceSend<P> = ForceSend(pointer);
+impl<T: Send + 'static> RcuDrop<T> {
+    /// Creates a new [`RcuDrop`] instance that delays the dropping of `value`.
+    pub fn new(value: T) -> Self {
+        Self {
+            value: ManuallyDrop::new(value),
+        }
+    }
+}
 
-    let rcu_monitor = RCU_MONITOR.get().unwrap();
-    rcu_monitor.after_grace_period(move || {
-        // This is necessary to make the Rust compiler to move the entire
-        // `ForceSend` structure into the closure.
-        let pointer = pointer;
+impl<T: Send + 'static> Deref for RcuDrop<T> {
+    type Target = T;
 
-        // SAFETY:
-        // 1. The pointer was previously returned by `into_raw`.
-        // 2. The pointer won't be used anymore since the grace period has
-        //    finished and this is the only time the pointer gets dropped.
-        let p = unsafe { <P as NonNullPtr>::from_raw(pointer.0) };
-        drop(p);
-    });
+    fn deref(&self) -> &Self::Target {
+        &self.value
+    }
+}
+
+impl<T: Send + 'static> DerefMut for RcuDrop<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.value
+    }
+}
```

Comment:
`DerefMut` is not sound. It's possible to drop the inner value with `core::mem::swap`. Don't add this. Fortunately, I don't think we really need this, right?